### ISSUE TITLE
substr should respect the UTF-8 flag

### ISSUE THREAD

**Issue by rdiez** [Perl/perl5#17897]
I have written a new file checksum tool in Perl:

[https://github.com/rdiez/Tools/tree/master/RDChecksum](https://github.com/rdiez/Tools/tree/master/RDChecksum)

During development, I noticed something with substr: I would have thought that doing a substr on a UTF-8 string would respect the UTF-8 flag. However, the string I get back is always flagged as "native/raw bytes" if it does not contain any high characters.

I started a discussion about it here:

https://www.perlmonks.org/?node_id=11118415

But nobody has yet given me a reason why substr should behave like that. It is surprising, because it is different to what happens if you capture a substring with a regular expression. I believe that the regex capture will be flagged as UTF-8 if the source string was also flagged with UTF-8, independently of whether there are high characters or not.

substr's behaviour could mean a performance hit later on, because Perl does not know anymore that the "native/raw bytes" string is safe for UTF-8, so it may decide to rescan it again when concatenating it to another UTF-8 string.

**Comment by Grinnz** [comment]
It is really difficult to guess what behavior you are describing without any example of what you're getting or what you think is flagging as "raw bytes" (Perl has no such mechanism). The UTF8 flag is not usable to user-facing code. When you get a string from any external input, unless it has gone through an encoding layer, it will be a string of bytes; whether it happens to be UTF-8 encoded text (or perhaps some other encoding) is up to you (or PerlIO layers) to determine by decoding it to characters if that is what you intend to work with. The same occurs with literal text in the source code; it is interpreted as the bytes in the file, unless `use utf8;` is active, whose only purpose is to cause any literal string to be interpreted as UTF-8-encoded text and result in the decoded characters instead. substr has no strange mechanism regarding strings; if you give it a string of bytes, you will get a substring of those bytes, and the same with characters.

So to summarize three basic points: The UTF8 flag does not indicate anything useful to your concerns here (many think that it is misnamed, my preferred way to refer to it is "upgraded string storage", as it does not indicate anything about the logical contents of the string). You must always know if you are working with a string of bytes or a string of characters. It is often best to decode text from UTF-8 directly after reading and encode it to UTF-8 directly before writing.

**Comment by rdiez** [comment]
Your points have already been expressed in the discussion linked above, where you will also find not only a confirmation of the described behaviour with detailed Perl code examples from other people, but also some notes and rebuttals from me.

In any case, I am still missing a justification about why substr does not behave like the rest of the string functions with regards to the UTF-8 flag. And yes, I still care about performance, even when writing Perl scripts. I worry that too many unnecessary string scans / conversions may be taking place behind the curtain.

**Comment by Leont** [comment]
99% of the time, if you feel the need to take a look at the UTF-8 flag you're doing the wrong thing. This has been explained to you already, repeatedly. Why open a bug here? Especially without actually describing a bug (e.g. with a code example).

**Comment by rdiez** [comment]
> 99% of the time, if you feel the need to take a look at the
> UTF-8 flag you're doing the wrong thing.
> This has been explained to you already, repeatedly.
> Why open a bug here?
> Especially without actually describing a bug (e.g. with a code example).

I am disappointed with your attitude.

The link above has code examples. Other people have immediately known what this is all about.

My script, linked above, is also code. Search for substr and you will find the comments about this issue.

I already said in the discussion I mentioned above that I am using that flag only for assertion purposes. My script explains how I am using it for that purpose. If you look inside, you will see that the flag is not used for any logic, only for assertions. Why do you categorically assume that I am wrong?

I raised a question about a possible performance issue about substr. Why do you ignore this matter? Can you name a reason why substr should behave like that anyway?

I already said that the vast majority of string routines follow the simple rule: UTF-8 in -> UTF-8 out. Regular expression capture does. chop does. Why substr should not?

**Comment by Grinnz** [comment]
The flag does not mean what you are asserting it to mean, so your assertions are misleading and your perception of the problem is wrong. I can attempt to explain it more fully but I have had this conversation so many times that it is quite tedious.

I am not aware of any differing behavior of substr, and I looked in the referenced thread and did not find any code examples to show either the bug or your misperception of such (misperception of these issues is extremely common around Unicode, and not to be ashamed of).

**Comment by rdiez** [comment]
Yes, I find discussing this subject with you rather tedious too. But look, I am not doing the wrong thing anymore, I'm just misled now. So we are actually moving forwards.  8-)

Just run this code:

<pre>
my $str = "Unicode character WHITE SMILING FACE: \x{263A}";
print "Step 1: " . utf8::is_utf8( $str ) . ", $str\n";
chop $str;
print "Step 2: " . utf8::is_utf8( $str ) . ", $str\n";
my @capture = $str =~ m/(Unicode character)/;
$str = $capture[0];
print "Step 3: " . utf8::is_utf8( $str ) . ", $str\n";
$str = lc( $str );
print "Step 4: " . utf8::is_utf8( $str ) . ", $str\n";
$str = reverse( $str );
print "Step 5: " . utf8::is_utf8( $str ) . ", $str\n";
$str =~ tr/c/x/;
print "Step 6: " . utf8::is_utf8( $str ) . ", $str\n";
$str = substr( $str, 0, 12 );
print "Step 7: " . utf8::is_utf8( $str ) . ", $str\n";
</pre>

You'll see this result:
<pre>
Wide character in print at ./test.pl line 5.
Step 1: 1, Unicode character WHITE SMILING FACE: ☺
Step 2: 1, Unicode character WHITE SMILING FACE: 
Step 3: 1, Unicode character
Step 4: 1, unicode character
Step 5: 1, retcarahc edocinu
Step 6: 1, retxarahx edoxinu
Step 7: , retxarahx ed
</pre>

Yes, the first warning "Wide character" is to be expected, just ignore it. Yes, the result is not wrong. Everything will be fine. We have talked about this before.

But still, look at step 7: the little "1" is missing! All other text manipulation routines "respect" the UTF-8 flag. Only substr does not.

Can you tell my why substr should be different?

Can you tell me whether this is not going to have performance implications later on? I mean, if you concatenate the string after calling substr() [with the UTF-8 flag turned off] with another string still flagged as UTF-8, what will happen?

Will Perl just concatenate the bytes?

Or will it not trust the string from substr anymore, because it has not got the UTF-8 flag set? Will it then scan the string looking for high characters, and try to convert them, or warn about them, during the concatenation?

**Comment by Grinnz** [comment]
Strings can be upgraded or downgraded at any time during normal operations. This has no bearing on further string operations. You've created a substring that no longer is required to be upgraded, so Perl has decided to downgrade its internal storage. You could manually upgrade that string or use it in an operation that invisibly upgrades it, and there would still be no difference in behavior of string operations.

Like I've been saying, the UTF8 bit does not mean what you think it means.

**Comment by Grinnz** [comment]
> Can you tell me whether this is not going to have performance implications later on? I mean, if you concatenate the string after calling substr() [with the UTF-8 flag turned off] with another string still flagged as UTF-8, what will happen?

For performance implications you would have to benchmark whatever particular case you are worried about, but I suspect it would not be significant. If you concatenate a downgraded string with an upgraded string, the result will usually be an upgraded string, unless it doesn't need to be (only contains codepoints under 255), then there is no guarantee either way.

**Comment by Grinnz** [comment]
As for the practical explanation of why only substr did that out of the operations you tried: the other operations are likely optimized to work in place in the way you used them, or at least not to do to the extra step of deciding whether the result could be downgraded. substr also has other concerns because its result can be used as an lvalue and it can replace text in place. But at the end of the day it doesn't affect behavior, it's an optimization. And it's very likely you'll find different results depending on Perl version by the way.

**Comment by rdiez** [comment]
> Like I've been saying, the UTF8 bit does not mean what you think it means.

Please stop saying that I misunderstand things. I have proved already that I know what I am talking about.

You did not answer my question about concatenation accurately. From your other vague explanations and suppositions, I think you do not actually know yourself, and cannot or will not actually find out in detail.

As a Perl outsider, I do admit that I do not know enough, but I will still venture the following supposition: this 'string downgrading' in substr is unnecessary and will hurt performance eventually. You may only notice if you process large strings. But you will eventually.

That is the main reason why I created this issue.

**Comment by Grinnz** [comment]
Okay, let's try starting over. Your framing of the issue indicated not understanding any of this, hence my responses.

The UTF8 bit does not indicate anything about the contents of the string. So let's stick with upgraded/downgraded string storage terminology.

The concern is that substr downgrades its result as a perhaps unnecessary optimization, while other string operations may not. As that simple statement, I think the issue is reasonable to consider.

**Comment by karenetheridge** [comment]
Regarding potentially inconsistent things happen when you append an upgraded string to a downgraded string, please see https://perldoc.pl/perlunicode#The-%22Unicode-Bug%22

In the perlmonks thread:
>  I am finding Unicode support in Perl hard. Most of my strings are ASCII, so there usually is no trouble. But then a Unicode character comes up, and suddenly writing text to stdout produces garbage characters and Perl issues a warning about it.
> So I have come up with an assert strategy: during development, I enable my "UTF-8 asserts", so that I verify that strings are flagged as native or as UTF-8 at the places where they should be. This has helped me prevent errors. And that is how I realised that substr() behaves differently. 

This is wrong. You cannot use the utf8 flag to track whether you have utf8 bytes or unicode characters. It is an internal flag used solely for optimizing internal operations, and as you have seen, upgrading or downgrading can happen at various times outside of your control.

You must keep track of your bytes-vs-characters status yourself, say by appending `_bytes` to your variable names.  The best practice here is to decode from bytes to characters as soon as possible when you receive data from a file or a terminal or the network, and always treat your strings as characters. Then, encode back into utf8 as late as possible when writing to a file or a terminal or the network.

**Comment by karenetheridge** [comment]
...but the perlmonks thread already contains lots of useful answers and advice, and unless there's an actual bug, this github queue is not really the best place to continue.

**Comment by rdiez** [comment]
> Okay, let's try starting over. Your framing of the issue indicated not understanding any of this, hence my responses.

My framing of the issue is fine. It is clear what I am talking about. The word "performance" is in my first post. It took a Perl hacker 12 minutes on perlmonks.org to confirm substr's odd behaviour. And the upgraded/downgraded terminology is actually misleading. We are not "upgrading" anything here.

> The concern is that substr downgrades its result as a perhaps unnecessary optimization, 

That is no "optimisation", it is exactly the opposite. If that were an optimisation, that would mean that all other string functions in my demo code above are not optimised, and therefore probably should be, shouldn't they?

I will state my theory again: If substr resets the UTF-8 flag, Perl will incur a performance penalty next time it needs to concatenate it with another string flagged as UTF-8. After all, if you are manipulating UTF-8 strings with substr and the like, the most likely scenario is that you will continue in the UTF-8 realm with the next string manipulations.

If you decide to encode/decode out of UTF-8, at least you know the points in your script where you will certainly incur a performance penalty. So, as far as I know, there is no good reason for substr to reset the string's UTF-8 flag, even if all characters are ASCII below 128.

> This is wrong. You cannot use the utf8 flag to track whether you have utf8 bytes or unicode characters. 

This is not wrong. Of course you can. It has helped me understand how Perl works and it has helped me locate the places in my script where I forgot to encode/decode to/from UTF-8. It has worked even when all characters were ASCII <  128. With the sole exception of substr.

It is fine as long as it is just an assertion strategy to be used during development, and as long as its limitations are clearly described. And it is indeed the case, for this is the comment I have in my script about it:

<pre>
# Do not enable these assertions for production, because Perl's internal behaviour may change and still
# remain compatible, breaking the checks but not really affecting functionality.
</pre>

Not only that, but this assertion strategy even looks like it has helped me find a performance issue in Perl. Or at least an oddity. Or at least something that probably should be clarified.

> You must keep track of your bytes-vs-characters status yourself, say by appending _bytes to your variable names. The best practice [...]

That is exactly what my script is doing, only that instead of appending "_bytes", I am appending "Utf8" to my string names. My script is fine. Would you please stop telling me about "best practice"? I am doing fine.

Just focus on this oddity in substr and its suspected performance implications. That is what this issue should be about.

> ...but the perlmonks thread already contains lots of useful answers and advice, and unless there's an actual bug, this github queue is not really the best place to continue.

This is not a bug, it is most likely a performance issue. That is why GitHub calls them "issues". Besides, anything that hurts performance in Perl bugs me!

And all that "useful answers and advice" has not yet  provided the answer to these simple questions: Is there a good reason why substr is resetting the UTF-8 flag? And what is the performance implications of that decision?
